modules/service/src/main/scala/lucuma/odb/service/ObservationService.scala

```diff
@@ -153,29 +172,110 @@ object ObservationService {
           pq.stream(which.argument, chunkSize = 1024).compile.toList
         }
 
+      // Select the observation ids from among the provided collection of ids which have an observing mode that does
+      // not match the `newMode`.
+      override def selectObservingModes(
+        which: List[Observation.Id]
+      ): F[Map[Option[ObservingModeType], List[Observation.Id]]] =
+        NonEmptyList
+          .fromList(which)
+          .fold(Applicative[F].pure(Map.empty)) { oids =>
+            val af = Statements.selectObservingModes(oids)
+            session.prepare(af.fragment.query(observation_id ~ observing_mode_type.opt)).use { pq =>
+              pq.stream(af.argument, chunkSize = 1024).compile.toList.map {
+                _.groupBy(_._2).view.mapValues(_.unzip._1).toMap
+              }
+            }
+          }
+
+      private def updateObservingModeType(
+        newMode: Option[ObservingModeType],
+        which:   List[Observation.Id]
+      ): F[Unit] = {
+        val af = Statements.updateObservingModeType(newMode, which)
+        session.prepare(af.fragment.command).use { pq =>
+          pq.execute(af.argument).void
+        }
+      }
+
+      private def updateObservingModes(
+        nEdit:           Nullable[ObservingModeInput.Edit],
+        rObservationIds: Result[List[Observation.Id]],
+        xa:              Transaction[F]
+      ): F[Result[List[Observation.Id]]] =
```

Comment:
This is maybe the most awkward part of the code.  Suggestions welcome.  It's tricky when the observing mode can be changed altogether, etc.